openedx/core/djangoapps/schedules/templatetags/ace.py

```diff
@@ -0,0 +1,150 @@
+
+    """
+    site, _user, message = _get_variables_from_context(context, 'with_link_tracking')
+
+    campaign = CampaignTrackingInfo(
+        source=message.app_label,
+        campaign=message.name,
+        content=message.uuid,
+    )
+    course_ids = context.get('course_ids')
+    if course_ids is not None and len(course_ids) > 0:
+        campaign.term = course_ids[0]
+
+    return mark_safe(
+        modify_url_to_track_clicks(
+            ensure_url_is_absolute(site, url),
+            campaign=campaign
+        )
+    )
+
+
+def _get_variables_from_context(context, tag_name):
+    if 'request' in context:
+        request = context['request']
+    else:
+        request = get_current_request()
+
+    if request is None:
+        raise template.VariableDoesNotExist(
+            'The {0} template tag requires a "request" to be present in the template context. Consider using '
+            '"emulate_http_request" if you are rendering the template in a celery task.'.format(tag_name)
+        )
+
+    message = context.get('message')
+    if message is None:
+        raise template.VariableDoesNotExist(
+            'The {0} template tag requires a "message" to be present in the template context.'.format(tag_name)
+        )
+
+    return request.site, request.user, message
+
+
+@register.simple_tag(takes_context=True)
+def google_analytics_tracking_pixel(context):
+    """
+    If configured, inject a google analytics tracking pixel into the template.
+
+    This tracking pixel will allow email open events to be tracked.
+
+    Args:
+        context (dict): The template context. Must include a "request" and "message".
+
+    Returns:
+        str: A string containing an HTML image tag that implements the GA measurement protocol or an empty string if
+             GA is not configured. For this to work, the site or settings must include the GA tracking ID.
+    """
+    image_url = _get_google_analytics_tracking_url(context)
+    if image_url is not None:
+        return mark_safe(
+            HTML('<img src="{0}" alt="" role="presentation" aria-hidden="true" />').format(HTML(image_url))
+        )
+    else:
+        return ''
+
+
+def _get_google_analytics_tracking_url(context):
+    site, user, message = _get_variables_from_context(context, 'google_analytics_tracking_pixel')
+
+    pixel = GoogleAnalyticsTrackingPixel(
+        site=site,
+        user_id=user.id,
+        campaign_source=message.app_label,
+        campaign_name=message.name,
+        campaign_content=message.uuid,
+        document_path='/email/{0}/{1}/{2}/{3}'.format(
+            message.app_label,
+            message.name,
+            message.send_uuid,
+            message.uuid,
+        ),
+    )
+    course_ids = context.get('course_ids')
+    if course_ids is not None and len(course_ids) > 0:
+        pixel.course_id = course_ids[0]
+
+    return pixel.image_url
```

Comment:
Ignorable Nit: Having `image_url` be a property hides the fact that there's actual computation behind it; the `course_id` better be set on the object before retrieving the `image_url`.  I wonder if it would be more comprehensible if it were a method instead.